ai_edge_torch/generative/layers/unet/blocks_2d.py

```diff
@@ -0,0 +1,183 @@
+    else:
+      self.residual_layer = nn.Conv2d(
+          config.in_channels, config.out_channels, kernel_size=1, stride=1, padding=0
+      )
+
+  def forward(
+      self, input_tensor: torch.Tensor, time_emb: Optional[torch.Tensor] = None
+  ) -> torch.Tensor:
+    residual = input_tensor
+    x = self.norm_1(input_tensor)
+    x = self.act_fn(x)
+    x = self.conv_1(x)
+    if self.time_emb_proj is not None:
+      time_emb = self.time_emb_proj(time_emb)[:, :, None, None]
+      x = x + time_emb
+    x = self.norm_2(x)
+    x = self.act_fn(x)
+    x = self.conv_2(x)
+    x = x + self.residual_layer(residual)
+    return x
+
+
+class AttentionBlock2D(nn.Module):
+  """2D self attention block with optional time embedding as input"""
+
+  def __init__(self, config: unet_cfg.AttentionBlock2DConfig):
+    super().__init__()
+    self.norm = layers_builder.build_norm(config.dims, config.normalization_config)
+    self.attention = SelfAttention(config.dims, config.attention_config, 0, True)
+
+  def forward(self, input_tensor: torch.Tensor) -> torch.Tensor:
+    residual = input_tensor
+    x = self.norm(input_tensor)
+    B, C, H, W = x.shape
+    x = x.view(B, C, H * W)
+    x = x.transpose(-1, -2)
+    x = self.attention(x)
+    x = x.transpose(-1, -2)
+    x = x.view(B, C, H, W)
+    x = x + residual
+    return x
+
+
+class UpDecoderBlock2D(nn.Module):
+  """Decoder block containing several residual blocks followed by an optional upsample"""
+
+  def __init__(self, config: unet_cfg.UpDecoderBlock2DConfig):
+    super().__init__()
+    self.config = config
+    resnets = []
+    for i in range(config.num_layers):
+      input_channels = config.in_channels if i == 0 else config.out_channels
+      resnets.append(
+          ResidualBlock2D(
+              unet_cfg.ResidualBlock2DConfig(
+                  in_channels=input_channels,
+                  out_channels=config.out_channels,
+                  time_embedding_channels=config.time_embedding_channels,
+                  normalization_config=config.normalization_config,
+                  activation_type=config.activation_type,
+              )
+          )
+      )
+    self.resnets = nn.ModuleList(resnets)
+    if config.add_upsample:
+      self.upsampler = unet_builder.build_upsampling(config.sampling_config)
+      if config.upsample_conv:
+        self.upsample_conv = nn.Conv2d(
+            config.out_channels, config.out_channels, kernel_size=3, stride=1, padding=1
+        )
+    else:
+      self.upsampler = None
+
+  def forward(
+      self, input_tensor: torch.Tensor, time_emb: Optional[torch.Tensor] = None
+  ) -> torch.Tensor:
+    hidden_states = input_tensor
+    for resnet in self.resnets:
+      hidden_states = resnet(hidden_states, time_emb)
+    if self.upsampler:
+      hidden_states = self.upsampler(hidden_states)
+      if self.upsample_conv:
+        hidden_states = self.upsample_conv(hidden_states)
+    return hidden_states
+
+
+class MidBlock2D(nn.Module):
```

Comment:
let's add more docstring / comments here to make the code easier to read.

Reply:
Done